hazelcast/src/main/java/com/hazelcast/core/Pipeline.java

```diff
@@ -0,0 +1,151 @@
+ * to prevent executing these requests. Invocations can be executed before the
+ * {@link #results()} is called.
+ *
+ * Pipelines can be used by both clients or members.
+ *
+ * The Pipeline isn't threadsafe. So only a single thread should add requests to
+ * the pipeline and wait for results.
+ *
+ * Currently all {@link ICompletableFuture} and their responses are stored in the
+ * Pipeline. So be careful executing a huge number of request with a single pipeline
+ * because it can lead to a huge memory bubble. In this cases it is better to
+ * periodically, after waiting for completion, to replace the pipeline by a new one.
+ * In the future we might provide this as an out of the box experience, but currently
+ * we do not.
+ *
+ * A pipeline provides its own backpressure on the system. So there will not be more
+ * in flight invocations than the depth of the pipeline. This means that the pipeline
+ * will work fine when backpressure on the client/member is disabled (default). Also
+ * when it it enabled it will work fine, but keep in mind that the number of concurrent
+ * invocations in the pipeline could be lower than the configured number of invocation
+ * of the pipeline because the backpressure on the client/member is leading.
+ *
+ * The Pipeline has been marked as Beta since we need to see how the API needs to
+ * evolve. But there is no problem using it in production. We use similar techniques
+ * to achieve high performance.
+ *
+ * @param <E>
+ */
+@Beta
+public class Pipeline<E> {
+
+    // we are using a caller runs executor to prevent having the overhead of creating a task and kicking an executor.
+    // the only thing that gets executed in future callback is the return of the license to the semaphore.
+    private static final Executor EXECUTOR = new Executor() {
+        @Override
+        public void execute(Runnable command) {
+            command.run();
+        }
+    };
+
+    private final Semaphore semaphore;
+    private final List<ICompletableFuture<E>> futures = new ArrayList<ICompletableFuture<E>>();
+
+    /**
+     * Creates a pipeline with the given depth.
+     *
+     * @param depth the maximum number of concurrent calls allowed in this pipeline.
+     * @throws IllegalArgumentException if depth smaller than 1. But if you use depth 1, it means that
+     *                                  every call is sync and you will not benefit from pipelining at all.
+     */
+    public Pipeline(int depth) {
+        checkPositive(depth, "depth must be positive");
+        this.semaphore = new Semaphore(depth);
+    }
+
+    /**
+     * Returns the results.
+     * <p>
+     * The results are returned in the order the requests were done.
+     * <p>
+     * This call waits till all requests have completed.
+     *
+     * @return the List of results.
+     * @throws Exception is something fails getting the results.
+     */
+    public List<E> results() throws Exception {
```

Comment:
Sorry for a little late call, but maybe it's a good idea to clear internal `futures` list when returning results? With this behavior a single pipeline objects can be used to process lots of data:
```java
Pipeline<String> pipeline = new Pipeline<String>(10);
for (int k = 0; k < 100; k++) {
    pipeline.add(map.getAsync(k));
    if (k % 20) {
        List<Integer> results = pipeline.results();
        // process intermediate results
    }
}
```

As it is now, it may be a bit misleading that a pipeline allows to add more operations after results were fetched and appends new results to the results list. With the current behavior I would prefer it to throw on `add()` once `results()` was called (to force users to avoid reusing pipeline objects).

Reply:
Good comments. That is why the API is beta. So being able to drain the pipeline instead of accumulating would be a nice out of the box feature. Best to see how this API is going to be used.

Also you probably want something like:
```
pipeline = new Pipeline(concurrent=100, unprocessed=1000)
    withExecutor(...).withConsumer(...)
```

So you can have the processing and draining of the pipeline as a direct feature instead of needing to add batching manually.